server/src/achievement/achievement.service.ts

```diff
@@ -1,17 +1,231 @@
-import { PrismaClient, User } from '@prisma/client';
+import {
+  $Enums,
+  Achievement,
+  AchievementTracker,
+  LocationType,
+  PrismaClient,
+  User,
+} from '@prisma/client';
 import { ClientService } from '../client/client.service';
 import { PrismaService } from '../prisma/prisma.service';
-import { AchievementDto, AchievementTrackerDto } from './achievement.dto';
+import { accessibleBy } from '@casl/prisma';
+import { AppAbility, CaslAbilityFactory } from '../casl/casl-ability.factory';
+import { Action } from '../casl/action.enum';
+import { subject } from '@casl/ability';
+import { defaultAchievementData } from '../organization/organization.service';
+import {
+  AchievementType,
+  AchievementDto,
+  LocationType as LocType,
+  AchievementTrackerDto,
+  UpdateAchievementDataDto,
+} from './achievement.dto';
 
 @Injectable()
 export class AchievementService {
   constructor(
-    private prisma: PrismaService,
+    private readonly prisma: PrismaService,
     private clientService: ClientService,
+    private abilityFactory: CaslAbilityFactory,
   ) {}
 
-  async getAwardFromId(id: string) {
+  /** get an achievement by its ID */
+  async getAchievementFromId(id: string) {
     return await this.prisma.achievement.findFirstOrThrow({ where: { id } });
   }
+
+  /** get list of achievements by IDs, based on ability */
+  async getAchievementsByIdsForAbility(
+    ability: AppAbility,
+    ids: string[],
+  ): Promise<Achievement[]> {
+    return await this.prisma.achievement.findMany({
+      where: {
+        AND: [
+          { id: { in: ids } },
+          accessibleBy(ability, Action.Read).Achievement,
+        ],
+      },
+    });
+  }
+
+  /** upsert achievement */
+  async upsertAchievementFromDto(
+    ability: AppAbility,
+    achievement: AchievementDto,
+  ) {
+    let ach = await this.prisma.achievement.findFirst({
+      where: { id: achievement.id },
+    });
+
+    if (
+      ach &&
+      (await this.prisma.achievement.findFirst({
+        select: { id: true },
+        where: {
+          AND: [
+            accessibleBy(ability, Action.Update).Achievement,
+            { id: ach.id },
+          ],
+        },
+      }))
+    ) {
+      const assignData = {
+        requiredPoints: achievement.requiredPoints,
+        name: achievement.name?.substring(0, 2048),
+        description: achievement.description?.substring(0, 2048),
+        imageUrl: achievement.imageUrl?.substring(0, 2048),
+        locationType: achievement.locationType as LocationType,
+        achievementType: achievement.achievementType as AchievementType,
+      };
+
+      const data = await this.abilityFactory.filterInaccessible(
+        assignData,
+        subject('Achievement', ach),
+        ability,
+        Action.Update,
+      );
+
+      ach = await this.prisma.achievement.update({
+        where: { id: ach.id },
+        data,
+      });
+    } else if (!ach && ability.can(Action.Create, 'Achievement')) {
```

Comment:
I think you should change the prisma schema to link achievements to organizations and check if this ability has the right to Action.Update the organization. Follow the initialOrganizationId thing that events have